### PR TITLE
Closes #24 Improve ChatSpamFilter

### LINKED ISSUE #24: Filter ChatSpam based on timestamp not ticks
Decouple the ticks logic of ais from the spam filter

### PR DESCRIPTION
**Pull request by cltnschlosser**
Closes #24

### CODE REVIEW

File: src/main/java/com/minecolonies/entity/ai/util/ChatSpamFilter.java [hunk 1]
```diff
@@ -17,37 +17,36 @@
      * 20 Ticks per second.
      * 30 Seconds.
      */
-    protected static final int BASE_TIMEOUT  = 30 * 20;
+    static final int BASE_TIMEOUT = 30 * 20;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/util/ChatSpamFilter.java]: CRITICAL Explicitly declare the visibility for "BASE_TIMEOUT". [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2039)

File: src/main/java/com/minecolonies/entity/ai/util/ChatSpamFilter.java [hunk 2]
```diff
@@ -17,37 +17,36 @@
...
      * 10 Minutes.
      */
-    protected static final int MAX_TIMEOUT   = 20 * 60 * 10;
-    /**
-     * The number to multiply timeout with every time
-     * <p>
-     * BASE_TIMEOUT * 2^times
-     */
-    private static final   int POWER_TIMEOUT = 2;
+    static final int MAX_TIMEOUT = 20 * 60 * 10;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/util/ChatSpamFilter.java]: CRITICAL Explicitly declare the visibility for "MAX_TIMEOUT". [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2039)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 2 issues
* CRITICAL 2 critical

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/entity/ai/util/ChatSpamFilter.java [hunk 3]
```diff
@@ -86,41 +74,31 @@ public void requestWithoutSpam(@NotNull String chat, int ticks)
...
-            if (speechDelay > 0)
+            if (speechDelay > worker.getOffsetTicks())
             {
-                speechDelay -= ticks;
                 return;
             }
-            speechRepeat++;
+
+            // to protect against overflows
+            if ((BASE_TIMEOUT << speechRepeat) < MAX_TIMEOUT)
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/entity/ai/util/ChatSpamFilter.java]: Can you add a comment explaining how this bitshift works and what it does exactly? That would greatly improve maintainability.

File: src/main/java/com/minecolonies/entity/ai/util/ChatSpamFilter.java [hunk 4]
```diff
@@ -86,41 +74,31 @@ public void requestWithoutSpam(@NotNull String chat, int ticks)
...
-     * It will be shown in certain delays.
-     * Helpful for requesting items.
-     *
-     * @param key  the translation key
-     * @param chat the chat message
-     */
-    public void talkWithoutSpam(String key, String... chat)
-    {
-        talkWithoutSpam(key, 1, chat);
+        speechDelay = Math.min(BASE_TIMEOUT << speechRepeat, MAX_TIMEOUT) + worker.getOffsetTicks();
```
Comment (Kostronor) [inline src/main/java/com/minecolonies/entity/ai/util/ChatSpamFilter.java]: Maybe for this bitshift too :)

File: src/test/java/com/minecolonies/entity/ai/util/ChatSpamFilterTest.java [hunk 5]
```diff
@@ -10,89 +13,98 @@
...
+    @Mock
+    private EntityCitizen citizen;
+
+    private ChatSpamFilter filter;
+
+    private static final String MESSAGE_1 = "Whatever";
+    private static final String MESSAGE_2 = "Whatever2";
+
+    @Before
+    public void setup() {
```
Comment (Kostronor) [inline src/test/java/com/minecolonies/entity/ai/util/ChatSpamFilterTest.java]: does the before resetup on each new test? Or do we need something like @BeforeEach?

Comment (cltnschlosser) [inline src/test/java/com/minecolonies/entity/ai/util/ChatSpamFilterTest.java]: Before each

**Comment by cltnschlosser** [comment]
:shipit:

**Comment by Kostronor** [comment]
:shipit: @Raycoms

**Comment by Raycoms** [comment]
:+1:

**Comment by cltnschlosser** [comment]
:shipit: why merge develop into this? @Kostronor You'll need to approve again

**Comment by cltnschlosser** [comment]
Why are we required to update the branch before we can merge?

**Pull request merged**